include/DDpackage.hpp

```diff
@@ -0,0 +1,2047 @@
+        }
+
+        CVec getVector(const vEdge& e) {
+            std::size_t dim = 1 << (e.p->v + 1);
+            // allocate resulting vector
+            auto vec = CVec(dim, {0.0, 0.0});
+            getVector(e, Complex::one, 0, vec);
+            return vec;
+        }
+        void getVector(const vEdge& e, const Complex& amp, std::size_t i, CVec& vec) {
+            // calculate new accumulated amplitude
+            auto c = cn.mulCached(e.w, amp);
+
+            // base case
+            if (e.isTerminal()) {
+                vec.at(i) = {c.r->val(), c.i->val()};
+                cn.returnToCache(c);
+                return;
+            }
+
+            std::size_t x = i | (1 << e.p->v);
+
+            // recursive case
+            if (!e.p->e[0].w.approximatelyZero())
+                getVector(e.p->e[0], c, i, vec);
+            if (!e.p->e[1].w.approximatelyZero())
+                getVector(e.p->e[1], c, x, vec);
+            cn.returnToCache(c);
+        }
+        void printVector(const vEdge& e) {
+            unsigned long long element = 2u << e.p->v;
+            for (unsigned long long i = 0; i < element; i++) {
+                auto amplitude = getValueByPath(e, i);
+                for (Qubit j = e.p->v; j >= 0; j--) {
+                    std::cout << ((i >> j) & 1u);
+                }
+                std::cout << ": " << amplitude << "\n";
+            }
+            std::cout << std::flush;
+        }
+
+        CMat getMatrix(const mEdge& e) {
+            std::size_t dim = 1 << (e.p->v + 1);
+            // allocate resulting matrix
+            auto mat = CMat(dim, CVec(dim, {0.0, 0.0}));
+            getMatrix(e, Complex::one, 0, 0, mat);
+            return mat;
+        }
+        void getMatrix(const mEdge& e, const Complex& amp, std::size_t i, std::size_t j, CMat& mat) {
+            // calculate new accumulated amplitude
+            auto c = cn.mulCached(e.w, amp);
+
+            // base case
+            if (e.isTerminal()) {
+                mat.at(i).at(j) = {c.r->val(), c.i->val()};
+                cn.returnToCache(c);
+                return;
+            }
+
+            std::size_t x = i | (1 << e.p->v);
+            std::size_t y = j | (1 << e.p->v);
+
+            // recursive case
+            if (!e.p->e[0].w.approximatelyZero())
+                getMatrix(e.p->e[0], c, i, j, mat);
+            if (!e.p->e[1].w.approximatelyZero())
+                getMatrix(e.p->e[1], c, i, y, mat);
+            if (!e.p->e[2].w.approximatelyZero())
+                getMatrix(e.p->e[2], c, x, j, mat);
+            if (!e.p->e[3].w.approximatelyZero())
+                getMatrix(e.p->e[3], c, x, y, mat);
+            cn.returnToCache(c);
+        }
+
+        ///
+        /// Deserialization
+        ///
+    public:
+        template<class Node, class Edge = Edge<Node>, std::size_t N = std::tuple_size_v<decltype(Node::e)>>
+        Edge deserialize(std::istream& is, bool readBinary = false) {
```

Comment:
Thinking about the deserialization, we should not use `int_fastX_t` so the width of fields does not change on different architectures. Maybe also add in documentation, that this is not suitable for longterm persistence?

This doesn't have to be done now, but we should keep it in the back of our heads.

Reply:
Changed it to `uint_least64_t` which should at least result in the same size of the serialised file on most 64bit systems.
I also added a note in the documentation indicating that de/serialization might not be portable.